### PR TITLE
Add minting algorithm to dposV3 contract

### PR DESCRIPTION
**Pull request by suhasagg**
- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [ ] All IP is original and not copied from another source
- [ ] I assign all copyright to Loom Network for the code in the pull request

Add minting algorithm to dposV3 contract
Adds new MintVouchers method to the DPOS contract that can be called only by delegators

### CODE REVIEW

File: builtin/plugins/coin/coin.go [hunk 1]
```diff
@@ -119,6 +119,22 @@ func (c *Coin) MintToGateway(ctx contract.Context, req *MintToGatewayRequest) er
 	return mint(ctx, gatewayAddr, &req.Amount.Value)
 }
 
+
+// MintToDPOS adds loom coins to the loom coin DPOS contract balance, and updates the total supply.
+func (c *Coin) MintToDPOS(ctx contract.Context, req *MintToDPOSRequest) error {
```
Comment (enlight) [inline builtin/plugins/coin/coin.go]: Don't need this, the DPOS contract will only mint ERC20

File: cmd/loom/dposV3_commands.go [hunk 2]
```diff
@@ -67,6 +67,68 @@ func UnjailValidatorCmdV3() *cobra.Command {
...
+	}
+	cli.AddContractCallFlags(cmd.Flags(), &flags)
+	return cmd
+}
+
+
+func MintERC20CmdV3() *cobra.Command {
+	var flags cli.ContractCallFlags
+	cmd := &cobra.Command{
+		Use:   "mint-erc20 [amount] [erc20token Address]",
```
Comment (enlight) [inline cmd/loom/dposV3_commands.go]: -> `mint-vouchers`

File: builtin/plugins/dposv3/dpos.go [hunk 3]
```diff
@@ -1521,6 +1582,12 @@ func loadCoin(ctx contract.Context) (*ERC20, error) {
 	}, nil
 }
 
+
+func loadERC20Token(ctx contract.Context, tokenAddr loom.Address) (*erc20Context, error) {
+	erc20 := newERC20Context(ctx, tokenAddr)
+    return erc20,nil
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: Formatting

File: builtin/plugins/dposv3/dpos.go [hunk 4]
```diff
@@ -292,6 +296,63 @@ func (c *DPOS) Delegate(ctx contract.Context, req *DelegateRequest) error {
 	return c.emitDelegatorDelegatesEvent(ctx, delegation)
 }
 
+//MintVouchers method to the DPOS contract that can be called only by delegators - Mints Loom Coin
+func (c *DPOS) MintVoucher(ctx contract.Context, req *MintVoucherRequest) error {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: Don't need this one.

File: builtin/plugins/dposv3/dpos.go [hunk 5]
```diff
@@ -292,6 +296,63 @@ func (c *DPOS) Delegate(ctx contract.Context, req *DelegateRequest) error {
...
+				return logDposError(ctx, err, transferFromErr)
+			}
+			break
+		}
+	}
+	return nil
+}
+
+//MintVouchersERC20 method to the DPOS contract that can be called only by delegators - Mints Generic ERC20 Token
+func (c *DPOS) MintVoucherERC20(ctx contract.Context, req *MintVoucherERC20Request) error {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: -> `MintVouchers`

File: builtin/plugins/dposv3/dpos.go [hunk 6]
```diff
@@ -292,6 +296,63 @@ func (c *DPOS) Delegate(ctx contract.Context, req *DelegateRequest) error {
...
+	delegations, err := loadDelegationList(ctx)
+	if err != nil {
+		return err
+	}
+	sender := ctx.Message().Sender
+	for _, d := range delegations {
+		if loom.UnmarshalAddressPB(d.Delegator).Compare(sender) == 0 {
+			erc20, err := loadERC20Token(ctx, loom.UnmarshalAddressPB(req.TokenAddress))
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: VoucherTokenAddress should be stored in `state.Params.VoucherTokenAddress`, not passed in the request.

File: builtin/plugins/dposv3/dpos.go (hunk 5 above)
Comment (enlight) [reply builtin/plugins/dposv3/dpos.go]: Check `req.Amount != nil` and greater than zero before doing anything else.

Comment (enlight) [reply builtin/plugins/dposv3/dpos.go]: And this new method should return an error if `dpos:v3.6` feature is not enabled.

File: builtin/plugins/coin/coin.go (hunk 1 above)
Comment (suhasagg) [reply builtin/plugins/coin/coin.go]: Done.

File: cmd/loom/dposV3_commands.go (hunk 2 above)
Comment (suhasagg) [reply cmd/loom/dposV3_commands.go]: Done.

File: builtin/plugins/dposv3/dpos.go (hunk 4 above)
Comment (suhasagg) [reply builtin/plugins/dposv3/dpos.go]: Done.

File: builtin/plugins/dposv3/dpos.go (hunk 3 above)
Comment (suhasagg) [reply builtin/plugins/dposv3/dpos.go]: Done.

File: builtin/plugins/dposv3/dpos.go (hunk 5 above)
Comment (suhasagg) [reply builtin/plugins/dposv3/dpos.go]: Added nil check for VoucherTokenAddress as well.

File: builtin/plugins/dposv3/dpos.go (hunk 6 above)
Comment (suhasagg) [reply builtin/plugins/dposv3/dpos.go]: Done

**Comment by suhasagg** [comment]
https://github.com/loomnetwork/loadtests/tree/MintableTokenDPOSContractTests
This branch deploys SampleERC20Token with minting support.
MinttoDOS function of evm contract is called to check erc20 balance of dpos contract is increasing with each mint command

**Comment by suhasagg** [comment]
New SampleERC20Token contract deployed with address:  default:0x7Fa7C465E138262f359FdE61E40aEE1837c2e2e7
Starting Contract calls test
MinttoDPOS
100
Balance Of
MinttoDPOS
200
Balance Of
MinttoDPOS
300
Balance Of
MinttoDPOS
400
Balance Of
MinttoDPOS
500
Balance Of
MinttoDPOS
600
Balance Of
MinttoDPOS
700
Balance Of
MinttoDPOS
800
Balance Of
MinttoDPOS
900
Balance Of
Completed Contract calls test
BalanceOf and minttoDPOS function called of erc20 token contract, as it can be seen with each minting dpos contract erc20 balance is increasing.

**Comment by suhasagg** [comment]
https://github.com/loomnetwork/go-loom/pull/426
Relevant go-loom PR

**Comment by suhasagg** [comment]
https://ci.kanwisher.com/blue/organizations/jenkins/loom-sdk-pipeline-prbuilder/detail/loom-sdk-pipeline-prbuilder/5138/pipeline/   Replayed build is building fine.

**Comment by enlight** [comment]
Will revisit this at a later date.

**Pull request closed without merging**